Python/specialize.c

```diff
@@ -2416,13 +2416,28 @@ binary_op_fail_kind(int oparg, PyObject *lhs, PyObject *rhs)
 
 /* float-long */
 
+// Guards should check that the float part is not NaN.
+// Guards for / (aka true_div) should check for 0 or 0.0 as the rhs.
 static int
 float_compactlong_guard(PyObject *lhs, PyObject *rhs)
 {
     return (
         PyFloat_CheckExact(lhs) &&
         PyLong_CheckExact(rhs) &&
-        _PyLong_IsCompact((PyLongObject *)rhs)
+        _PyLong_IsCompact((PyLongObject *)rhs) &&
+        !isnan(PyFloat_AsDouble(lhs))
+    );
+}
+
+static int
+float_compactlong_guard_true_div(PyObject *lhs, PyObject *rhs)
+{
+    return (
+        PyFloat_CheckExact(lhs) &&
+        PyLong_CheckExact(rhs) &&
+        _PyLong_IsCompact((PyLongObject *)rhs) &&
+        !PyLong_IsZero(rhs) &&
+        !isnan(PyFloat_AsDouble(lhs))
```

Comment:
Except for `!PyLong_IsZero(rhs)`, this seems to be the same as `float_compactlong_guard`. Maybe you could reuse it to avoid some duplication?

```suggestion
        float_compactlong_guard(lhs, rhs) && !PyLong_IsZero(rhs)
```

Reply:
You're right, Tomas. Also the same technique can be used for the `compactlong_float_guard_true_div`:
`compactlong_float_guard(lhs, rhs) && PyFloat_AsDouble(rhs) != 0.0`. I also mark these guards with `inline` keyword.